File: reference/python/hbprep/palette.py
# -*- coding: utf-8 -*-
"""Farbvergabe pro Figur.

Zwei Ebenen, damit ein Kapitel nie zwei Figuren in derselben Farbe zeigt und
die Hauptfiguren trotzdem im ganzen Buch dieselbe Farbe behalten:

  * Hauptcast  -> global feste Slots (buchweit stabil)
  * Nebenfiguren -> Slot pro Kapitel, kollisionsfrei recycelt

Redundanz für Farbfehlsichtige und Schwarz-Weiß-Druck: jede Figur trägt ein
Initialen-Badge, im Unterstreichungsmodus hat jeder Slot zusätzlich ein Muster.
"""

from __future__ import annotations

# Textmarker-Palette. Der Text bleibt schwarz bzw. hell, eingefärbt wird nur
# das Markerband:
#   tl / td  – Markerband im hellen / dunklen Thema
#   l  / d   – kräftiger Ton derselben Farbe (Badge, Legendenrand, Unterstreichung)
# Reihenfolge = Vergabereihenfolge: die am besten unterscheidbaren zuerst.
SLOTS = [
    {"name": "Gelb",    "l": "#8a6d00", "d": "#f2d64b", "tl": "#fff27a", "td": "#4d4300"},
    {"name": "Grün",    "l": "#2e7d32", "d": "#86d98a", "tl": "#bff5b0", "td": "#1e4520"},
    {"name": "Pink",    "l": "#b0276f", "d": "#ff94cb", "tl": "#ffc6e6", "td": "#561c3b"},
    {"name": "Blau",    "l": "#1f5fb8", "d": "#8cc4ff", "tl": "#b8dfff", "td": "#173a63"},
    {"name": "Orange",  "l": "#b35900", "d": "#ffb066", "tl": "#ffd5a1", "td": "#5a3310"},
    {"name": "Lila",    "l": "#6a3fc2", "d": "#c0a3ff", "tl": "#ddcbff", "td": "#35265e"},
    {"name": "Türkis",  "l": "#0b7a6e", "d": "#66dccb", "tl": "#aef1e6", "td": "#124640"},
    {"name": "Limette", "l": "#5c7a00", "d": "#cbe45f", "tl": "#e6f7a0", "td": "#3a4812"},
    {"name": "Koralle", "l": "#b3261e", "d": "#ff958b", "tl": "#ffc0b8", "td": "#57201c"},
    {"name": "Himmel",  "l": "#3f51b5", "d": "#a5b4ff", "tl": "#cfd8ff", "td": "#27305a"},
    {"name": "Sand",    "l": "#7a5c1f", "d": "#dcc38c", "tl": "#eedfb9", "td": "#473a22"},
    {"name": "Mauve",   "l": "#8a4a7a", "d": "#e0a8d2", "tl": "#ecd0e5", "td": "#4a2c43"},
]
PATTERNS = ["solid", "dashed", "dotted", "double"]
MISC = {"name": "Neben", "l": "#5c6470", "d": "#a4adba", "tl": "#e2e5ea", "td": "#2c323a"}


def initials(name: str) -> str:
    parts = [p for p in name.replace("-", " ").split() if p[:1].isalpha()]
    if not parts:
        return "?"
    if len(parts) == 1:
        return parts[0][:2].capitalize()
    return (parts[0][0] + parts[-1][0]).upper()


def assign(doc: dict, main_min_lines: int = 4, max_main: int = 12) -> dict:
    cast = doc.get("cast", [])
    by_id = {c["id"]: c for c in cast}

    # 1) Hauptcast – nach Redeanteil, respektiert vorgegebene Farben aus cast.json.
    #    Doppelt vergebene Slots werden aufgelöst: die redestärkere Figur behält
    #    ihre Farbe, die andere rückt auf eine freie.
    forced = sorted((c for c in cast if isinstance(c.get("color"), int)),
                    key=lambda c: -c["lines"])
    used: set[int] = set()
    for c in forced:
        want = c["color"]
        if 0 <= want < len(SLOTS) and want not in used:
            c["slot"] = want
            used.add(want)
        else:
            c["slot"] = None

    rest = [c for c in cast if c["lines"] >= main_min_lines
            and not isinstance(c.get("color"), int)]
    main = list(forced) + rest[:max(0, max_main - len(forced))]

    free = [i for i in range(len(SLOTS)) if i not in used]
    for c in main:
        if c.get("slot") is None:
            c["slot"] = free.pop(0) if free else None
        c["badge"] = c.get("badge") or initials(c["name"])
    for c in cast:
        c.setdefault("slot", None)
        c["badge"] = c.get("badge") or initials(c["name"])
        c["color"] = c["slot"]

    # 2) Nebenfiguren – Slot pro Kapitel, ohne Kollision
    ch_colors: dict[str, dict[str, int]] = {}
    for ch in doc["chapters"]:
        present, order = set(), []
        for b in ch["blocks"]:
            for sp in b.get("speech", []):
                cid = sp.get("speaker")
                if cid and cid not in present:
                    present.add(cid)
                    order.append(cid)
        taken = {by_id[c]["slot"] for c in present
                 if c in by_id and by_id[c]["slot"] is not None}
        local, pool = {}, [i for i in range(len(SLOTS)) if i not in taken]
        for cid in order:
            c = by_id.get(cid)
            if not c or c["slot"] is not None:
                continue
            local[cid] = pool.pop(0) if pool else None
        ch_colors[ch["id"]] = local

    doc["chapter_colors"] = ch_colors
    doc["palette"] = {"slots": SLOTS, "patterns": PATTERNS, "misc": MISC}
    return doc


def slot_for(doc: dict, chapter_id: str, cast_id: str | None) -> int | None:
    if not cast_id:
        return None
    for c in doc.get("cast", []):
        if c["id"] == cast_id and c.get("slot") is not None:
            return c["slot"]
    return doc.get("chapter_colors", {}).get(chapter_id, {}).get(cast_id)

File: packages/core/src/pipeline/palette.ts
/**
 * Farbvergabe pro Figur (Textmarker).
 *
 * Zwei Ebenen: Der Hauptcast bekommt buchweit feste Slots; Nebenfiguren
 * bekommen pro Kapitel einen dort freien Slot. So ist innerhalb eines Kapitels
 * keine Farbe doppelt vergeben, und Hauptfiguren behalten ihre Farbe im ganzen
 * Buch. Redundanz zur Farbe liefert das Initialen-Kürzel.
 */
import { capitalize, splitWs } from "../text.js";
import type { CastMember, Doc } from "./types.js";

export interface MarkerSlot {
  name: string;
  /** Markerband im hellen Thema */
  light: string;
  /** Markerband im dunklen Thema */
  dark: string;
  /** Kräftiger Ton für Kürzel und Ränder (hell / dunkel) */
  strongLight: string;
  strongDark: string;
}

/** Reihenfolge = Vergabereihenfolge: die am besten unterscheidbaren zuerst. */
export const MARKER_SLOTS: readonly MarkerSlot[] = [
  { name: "Gelb", light: "#fff27a", dark: "#4d4300", strongLight: "#8a6d00", strongDark: "#f2d64b" },
  { name: "Grün", light: "#bff5b0", dark: "#1e4520", strongLight: "#2e7d32", strongDark: "#86d98a" },
  { name: "Pink", light: "#ffc6e6", dark: "#561c3b", strongLight: "#b0276f", strongDark: "#ff94cb" },
  { name: "Blau", light: "#b8dfff", dark: "#173a63", strongLight: "#1f5fb8", strongDark: "#8cc4ff" },
  { name: "Orange", light: "#ffd5a1", dark: "#5a3310", strongLight: "#b35900", strongDark: "#ffb066" },
  { name: "Lila", light: "#ddcbff", dark: "#35265e", strongLight: "#6a3fc2", strongDark: "#c0a3ff" },
  { name: "Türkis", light: "#aef1e6", dark: "#124640", strongLight: "#0b7a6e", strongDark: "#66dccb" },
  { name: "Limette", light: "#e6f7a0", dark: "#3a4812", strongLight: "#5c7a00", strongDark: "#cbe45f" },
  { name: "Koralle", light: "#ffc0b8", dark: "#57201c", strongLight: "#b3261e", strongDark: "#ff958b" },
  { name: "Himmel", light: "#cfd8ff", dark: "#27305a", strongLight: "#3f51b5", strongDark: "#a5b4ff" },
  { name: "Sand", light: "#eedfb9", dark: "#473a22", strongLight: "#7a5c1f", strongDark: "#dcc38c" },
  { name: "Mauve", light: "#ecd0e5", dark: "#4a2c43", strongLight: "#8a4a7a", strongDark: "#e0a8d2" },
];

export const MARKER_MISC: MarkerSlot = {
  name: "Neben", light: "#e2e5ea", dark: "#2c323a", strongLight: "#5c6470", strongDark: "#a4adba",
};

/**
 * Stiftfarben für Betonungen. Bewusst kräftige Linientöne statt der Markerbänder der Figuren, und jede
 * Farbe mit eigener Linienart – so bleiben Betonungen auch ohne Farbe unterscheidbar (Studio-Thema,
 * Farbsehschwäche, Schwarzweißdruck). Gewellt in Rot wäre ein Retake, deshalb ist Rot doppelt.
 */
export interface PenSlot {
  name: string;
  line: "double" | "wavy" | "dotted" | "dashed" | "solid";
  /** Linienfarbe im hellen / dunklen Thema */
  light: string;
  dark: string;
}

export const PEN_SLOTS: readonly PenSlot[] = [
  { name: "Rot", line: "double", light: "#c62828", dark: "#ff8a80" },
  { name: "Blau", line: "wavy", light: "#1565c0", dark: "#82b1ff" },
  { name: "Grün", line: "dotted", light: "#2e7d32", dark: "#69f0ae" },
  { name: "Orange", line: "dashed", light: "#d84315", dark: "#ffab40" },
  { name: "Violett", line: "solid", light: "#6a1b9a", dark: "#ea80fc" },
];

/** Stiftfarbe einer Betonung – unbekannte Indizes (neuere Version) gelten als schlicht. */
export const penSlot = (color: number | null | undefined): number | null =>
  color !== null && color !== undefined && Number.isInteger(color) && color >= 0 && color < PEN_SLOTS.length ? color : null;

export function initials(name: string): string {
  const parts = splitWs(name.replace(/-/g, " ")).filter((p) => /^\p{L}/u.test(p));
  if (!parts.length) return "?";
  if (parts.length === 1) return capitalize(parts[0]!.slice(0, 2));
  return (parts[0]![0]! + parts.at(-1)![0]!).toUpperCase();
}

export interface PaletteOptions {
  /** Ab so vielen Redeteilen bekommt eine Figur eine buchweit feste Farbe. */
  mainMinLines?: number;
  maxMain?: number;
}

const isSlot = (v: unknown): v is number => typeof v === "number" && Number.isInteger(v);

export function assignColors(doc: Doc, opts: PaletteOptions = {}): Doc {
  const mainMinLines = opts.mainMinLines ?? 4;
  const maxMain = opts.maxMain ?? MARKER_SLOTS.length;
  const cast: CastMember[] = doc.cast ?? [];
  const byId = new Map(cast.map((c) => [c.id, c]));

  // 1) Hauptcast. Vorgaben werden respektiert; doppelte Vorgaben löst die
  //    redestärkere Figur für sich, die andere rückt auf einen freien Slot.
  const forced = cast.filter((c) => isSlot(c.color)).sort((a, b) => b.lines - a.lines);
  const used = new Set<number>();
  for (const c of forced) {
    const want = c.color as number;
    if (want >= 0 && want < MARKER_SLOTS.length && !used.has(want)) {
      c.slot = want;
      used.add(want);
    } else {
      c.slot = null;
    }
  }
  const rest = cast.filter((c) => c.lines >= mainMinLines && !isSlot(c.color));
  const main = [...forced, ...rest.slice(0, Math.max(0, maxMain - forced.length))];
  const free = MARKER_SLOTS.map((_, i) => i).filter((i) => !used.has(i));
  for (const c of main) {
    if (c.slot === undefined || c.slot === null) c.slot = free.shift() ?? null;
    c.badge ||= initials(c.name);
  }
  for (const c of cast) {
    c.slot ??= null;
    c.badge ||= initials(c.name);
    c.color = c.slot;
  }

  // 2) Nebenfiguren – ein Slot pro Kapitel, ohne Kollision
  const chapterColors: Record<string, Record<string, number | null>> = {};
  for (const ch of doc.chapters) {
    const order: string[] = [];
    for (const b of ch.blocks) {
      for (const sp of b.speech ?? []) {
        if (sp.speaker && !order.includes(sp.speaker)) order.push(sp.speaker);
      }
    }
    const taken = new Set(order.map((id) => byId.get(id)?.slot).filter(isSlot));
    const pool = MARKER_SLOTS.map((_, i) => i).filter((i) => !taken.has(i));
    const local: Record<string, number | null> = {};
    for (const id of order) {
      const c = byId.get(id);
      if (!c || c.slot !== null) continue;
      local[id] = pool.shift() ?? null;
    }
    chapterColors[ch.id] = local;
  }
  doc.chapter_colors = chapterColors;
  return doc;
}
